allennlp/models/encoder_decoders/encoder_decoder.py

```diff
@@ -0,0 +1,274 @@
+        self._decoder_output_dim = self._encoder.get_output_dim()
+        # TODO (pradeep): target_embedding_dim need not be the same as the source embedding dim.
+        target_embedding_dim = self._source_embedder.get_output_dim()
+        self._target_embedder = Embedding(self._num_classes, target_embedding_dim)
+        if self._use_attention:
+            self._decoder_attention = Attention()
+            # The output of attention will be concatenated to the input vector of the decoder at
+            # each time step.
+            self._decoder_input_dim = self._encoder.get_output_dim() + target_embedding_dim
+        else:
+            self._decoder_input_dim = target_embedding_dim
+        # TODO (pradeep): Do not hardcode decoder cell type.
+        self._decoder_cell = LSTMCell(self._decoder_input_dim, self._decoder_output_dim)
+        self._output_projection_layer = Linear(self._decoder_output_dim, self._num_classes)
+        # We need the start symbol to provide as the input at the first timestep of decoding, and end symbol
+        # as a way of letting decoder choose to stop decoding.
+
+    @overrides
+    def forward(self,  # type: ignore
+                source_tokens: Dict[str, torch.LongTensor],
+                target_tokens: Dict[str, torch.LongTensor] = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        """
+        Decoder logic for producing the entire target sequence.
+
+        Parameters
+        ----------
+        source_tokens : Dict[str, torch.LongTensor]
+           The output of ``TextField.as_array()`` applied on the input ``TextField``. This will be passed
+           through a ``TextFieldEmbedder`` and then through an encoder.
+        target_tokens : Dict[str, torch.LongTensor], optional (default = None)
+            Dict with indices of target side tokens, of shape ``(batch_size, target_sequence_length)``
+            in ``tokens`` field.
+        """
+        # (batch_size, input_seq_length, enc_output_dim)
+        encoder_outputs = self._get_encoder_outputs(source_tokens)
+        final_encoder_output = encoder_outputs[:, -1]  # (batch_size, enc_output_dim)
+        if target_tokens:
+            targets = target_tokens["tokens"]
+            target_sequence_length = targets.size()[1]
+            num_decoding_steps = target_sequence_length
+        else:
+            num_decoding_steps = self._max_decoding_steps
+        # TODO (pradeep): Define a DecoderState class?
+        decoder_hidden = final_encoder_output
+        with torch.cuda.device_of(decoder_hidden):
+            decoder_context = Variable(torch.zeros(1, self._decoder_output_dim))
+        last_predictions = None
+        step_logits = []
+        step_probabilities = []
+        step_predictions = []
+        for timestep in range(num_decoding_steps):
+            if self.training and all(torch.rand(1) >= self._scheduled_sampling_ratio):
```

Comment:
Hmm.. we'll need to gather appropriate rows from predictions and targets and reconstruct a batch. Not sure if we'll actually gain anything by doing that.

Reply:
You don't need a gather, you can do it reasonably efficiently by doing something like:

```python
sampling_mask = torch.rand(batch_size) >= self._scheduled_sampling_ratio
forcing_input = targets[:, timestep]
sampling_input = ...  # what's currently in the `else` clause
input_choices = sampling_mask * forcing_input + (1 - sampling_mask)  * sampling_input
```

This would do a better job of actually mixing according to the sampling schedule you set.